Tools/cases_generator/analyzer.py

```diff
@@ -0,0 +1,410 @@
+    type: str | None
+    condition: str | None
+    size: str
+    peek: bool = False
+
+    def __str__(self):
+        cond = " if (" + self.condition + ")" if self.condition else ""
+        size = "[" + self.size + "]" if self.size != "1" else ""
+        type = "" if self.type is None else self.type + " "
+        return f"{type}{self.name}{size}{cond} {self.peek}"
+
+    def is_array(self):
+        return self.type == "PyObject **"
+
+@dataclass
+class StackEffect:
+    inputs: list[StackItem]
+    outputs: list[StackItem]
+
+    def __str__(self):
+        return f"({', '.join([str(i) for i in self.inputs])} -- {', '.join([str(i) for i in self.outputs])})"
+
+@dataclass
+class CacheEntry:
+    name: str
+    size: int
+
+    def __str__(self):
+        return f"{self.name}/{self.size}"
+
+@dataclass
+class Uop:
+    name: str
+    context: parser.Context
+    annotations: list[str]
+    stack: StackEffect
+    caches: list[CacheEntry]
+    body: list[lexer.Token]
+    properties: Properties
+    _size: int = -1
+
+    def dump(self, indent):
+        print(indent, self.name, ", ".join(self.annotations) if self.annotations else "")
+        print(indent, self.stack, ", ".join([str(c) for c in self.caches]))
+        self.properties.dump("    " + indent)
+
+    @property
+    def size(self):
+        if self._size < 0:
+            self._size = sum(c.size for c in self.caches)
+        return self._size
+
+Part = Uop | Skip
+
+@dataclass
+class Instruction:
+    name: str
+    uops: list[Part]
+    _properties : Properties | None
+    is_target: bool = False
+    family: "Family" = None
+
+    @property
+    def properties(self) -> Properties:
+        if self._properties is None:
+            self._properties = self._compute_properties()
+        return self._properties
+
+    def _compute_properties(self) -> Properties:
+        return Properties(
+            any(part.properties.escapes for part in self.uops),
+            all(part.properties.infallible for part in self.uops),
+            any(part.properties.deopts for part in self.uops),
+            any(part.properties.oparg for part in self.uops),
+            any(part.properties.jumps for part in self.uops),
+            any(part.properties.ends_with_eval_breaker for part in self.uops),
+            any(part.properties.needs_this for part in self.uops),
+            any(part.properties.always_exits for part in self.uops),
+            any(part.properties.stores_sp for part in self.uops),
+        )
+
+    def dump(self, indent):
+        print(indent, self.name, "=", ", ".join([op.name for op in self.uops]))
+        self.properties.dump("    " + indent)
+
+    @property
+    def size(self):
+        return 1 + sum(uop.size for uop in self.uops)
+
+@dataclass
+class PseudoInstruction:
+    name: str
+    targets: list[Instruction]
+
+    def dump(self, indent):
+        print(indent, self.name, "->", " or ".join([t.name for t in self.targets]))
+
+
+@dataclass
+class Family:
```

Comment:
This reminds me of another possible naming scheme. We already have `parser.InstDef` vs. `Instruction`. Maybe the other top-level things that the parser recognizes could also be named `FooDef`? I.e., `FamilyDef`, `PseudoDef`, `MacroDef`.

Reply:
That sounds like a good scheme.